---
title: "ENCODE DREAM Challenge"
author: John Reid
output: pdf_document
---

```{r echo=FALSE, eval=FALSE}
#
# Execute this block to render this document.
#
devtools::load_all()
rmarkdown::render('ENCODE-DREAM-overview.Rmd')
```


```{r parseCombos, echo=FALSE, message=FALSE}
library(Saturn)
library(stringr)
library(dplyr)
library(reshape2)
library(ggplot2)
library(ggthemes)
#
# Parse the information about the TFs.
scraped <-
'ARID3A 	HepG2 	K562 	
ATF2 	GM12878, H1-hESC, MCF-7 	K562 	HepG2
ATF3 	HCT116, H1-hESC, HepG2, K562 	liver 	
ATF7 	GM12878, HepG2, K562 	MCF-7 	
CEBPB 	A549, H1-hESC, HCT116, HeLa-S3, HepG2, IMR-90, K562 	MCF-7 	
CREB1 	GM12878, H1-hESC, HepG2, K562 	MCF-7 	
CTCF 	A549, H1-hESC, HeLa-S3, HepG2, IMR-90, K562, MCF-7 	GM12878 	PC-3, induced_pluripotent_stem_cell
E2F1 	GM12878, HeLa-S3 		K562
E2F6 	A549, H1-hESC, HeLa-S3 	K562 	
EGR1 	GM12878, H1-hESC, HCT116, MCF-7 	K562 	liver
EP300 	GM12878, H1-hESC, HeLa-S3, HepG2, K562, SK-N-SH 	MCF-7 	
FOXA1 	HepG2 	MCF-7 	liver
FOXA2 	HepG2 		liver
GABPA 	GM12878, H1-hESC, HeLa-S3, HepG2, MCF-7, SK-N-SH 	K562 	liver
GATA3 	A549, SK-N-SH 	MCF-7 	
HNF4A 	HepG2 		liver
JUND 	HCT116, HeLa-S3, HepG2, K562, MCF-7, SK-N-SH 	H1-hESC 	liver
MAFK 	GM12878, H1-hESC, HeLa-S3, HepG2, IMR-90 	K562, MCF-7 	
MAX 	A549, GM12878, H1-hESC, HCT116, HeLa-S3, HepG2, K562, SK-N-SH 	MCF-7 	liver
MYC 	A549, HeLa-S3, K562, MCF-7 	HepG2 	
NANOG 	H1-hESC 		induced_pluripotent_stem_cell
REST 	H1-hESC, HeLa-S3, HepG2, MCF-7, Panc1, SK-N-SH 	K562 	liver
RFX5 	GM12878, HeLa-S3, MCF-7, SK-N-SH 	HepG2 	
SPI1 	GM12878 	K562 	
SRF 	GM12878, H1-hESC, HCT116, HepG2, K562 	MCF-7 	
STAT3 	HeLa-S3 	GM12878 	
TAF1 	GM12878, H1-hESC, HeLa-S3, K562, SK-N-SH 	HepG2 	liver
TCF12 	GM12878, H1-hESC, MCF-7, SK-N-SH 	K562 	
TCF7L2 	HCT116, HeLa-S3, Panc1 	MCF-7 	
TEAD4 	A549, H1-hESC, HCT116, HepG2, K562, SK-N-SH 	MCF-7 	
YY1 	GM12878, H1-hESC, HCT116, HepG2, SK-N-SH 	K562 	
ZNF143 	GM12878, H1-hESC, HeLa-S3, HepG2 	K562 	'
lines <- str_split(scraped, fixed('\n'))[[1]]
parse.line <- function(l) {
    # print(l)
    parsed <- str_trim(str_split(l, fixed('\t'), 4)[[1]])
    # parsed[[2]] <- str_split(parsed[[2]], fixed(', '))[[1]]
    # print(parsed)
    parsed
}
parsed <- lapply(lines, parse.line)
split.for.col <- function(col) {
    if (2 == col) return('train')
    if (3 == col) return('ladder')
    if (4 == col) return('submit')
}
parse.tfs <- function(parsed, col) {
    data.frame(
        TF=parsed[[1]],
        cell=str_split(parsed[[col]], fixed(', '))[[1]],
        split=split.for.col(col))
}
parse.cols <- function(parsed) {
    res <- parse.tfs(parsed, 2)
    if (length(parsed) > 2) res <- rbind(res, parse.tfs(parsed, 3))
    if (length(parsed) > 3) res <- rbind(res, parse.tfs(parsed, 4))
    res %>% filter("" != cell)
}
tfs <- do.call(rbind, lapply(parsed, parse.cols))
tf.levels <- levels(tfs$TF)
cell.levels <- levels(tfs$cell)
#
# Add information about which TF/cell type combos are true blind
true.blind.ladder <- data.frame(
    TF = factor(c('ATF2', 'ATF7', 'CREB1', 'FOXA1', 'TCF12'), levels=tf.levels),
    cell = factor(c('K562', 'MCF-7', 'MCF-7', 'MCF-7', 'K562'), levels=cell.levels),
    true.blind = TRUE)
tfs <- tfs %>% left_join(true.blind.ladder) %>%
    mutate(
        true.blind = ifelse(is.na(true.blind), 'submit' == split, true.blind),
        split.blind = ifelse('ladder'==split,
                             ifelse(true.blind, 'ladder.blind', 'ladder'),
                             as.character(split)))
```


# The ENCODE-DREAM challenge

The [challenge](https://www.synapse.org/#!Synapse:syn6131484/wiki/) is to
predict cell-type specific binding of transcription factors (TFs) using four
types of data:

- DNA sequence: a reference human genomic sequence
- DNA shape: the physical shape of the genome
- DNase-seq: experimental evidence for how open and accessible regions of the genome are
- RNA-seq: gene expression levels


## TF/cell-type combinations

TFs have different binding profiles in different cell types. `r length(unique(tfs$TF))`
TFs and `r length(unique(tfs$cell))` cell types are represented in the data although not
all combinations are present.

```{r, echo=FALSE}
ggplot(tfs, aes(TF, cell)) +
    geom_tile(aes(fill=split.blind), colour='white') +
    scale_fill_few() +
    theme_tufte() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=.4))
```

There are essentially three (two!?) prediction tasks:

- Held-out chromosomes: For each TF/cell-type combination, the ChIP-seq data will
  not be available for 3 chromosomes. Methods will be assessed by their predictive
  performance on these chromosomes.
- Across cell-type: Methods will be assessed by their performance on cell-types for
  which no training data has been made available (PC-3, induced pluripotent stem
  cells, liver).
- Within cell-type: Methods will be assessed on 


## Data


### TF binding

TF binding is measured using the ChIP-seq protocol and converted to a binary
score for sliding windows of 200bp. The windows slide by 50bp. For each 200bp
location, binding is defined as bound (B), unbound (U) or ambiguous (A). In
addition the challenge provides more detailed information from the ChIP-seq
experiments including conservative and relaxed estimates of peaks and
fold-control signals showing how enriched the ChIP-seq experiment was over a
control background experiment. It is not clear how useful the extra information
will be as it will obviously not be available on the held-out data.


### DNA sequence

The human reference genome is over 3 billion base pairs long. Each base is
represented as a character: adenine (A); cytosine (C); guanine (G) and thymine
(T). The challenge is restricted to chromosomes 1-22 and chromosome X. All data
in the challenge are defined with respect to release GRCh37/hg19.  TFs tend to
prefer to bind specific sequences. These preferences are summarised in binding
motifs but these are not known for all TFs. The only external data that is
allowed to be used in this challenge are libraries of TF-DNA binding motifs.
Obviously this data is not cell-type specific.


### DNase-seq

Information on chromatin accessibility on a per-cell-type basis will be
available in four formats:

- conservative peaks
- relaxed peaks
- filtered BAM alignment files
- fold-enrichment signal coverage tracks

The first two will be much easier to use as they summarise the last two.


### DNA shape

Participants are encouraged to use the [DNAshapeR](http://tsupeichiu.github.io/DNAshapeR/) to
calculate DNA shape features across the genome. This information is not cell-type specific but
has been shown to be predictive of TF-DNA binding. Note that `DNAshapeR` requires a version of
R $> 3.3$.


### Gene expression

Gene expression is regulated by TF binding and so should be useful indirectly to infer binding.
The major difficulties are that
- It is known which locations on the genome regulate which genes. Commonly TF binding sites
  regulate the closest gene but this is not always the case.
- Several TFs can combine in an unknown way to regulate a gene.
- There are other mechanisms of gene regulation that will confound the relationship.



## Transcription factors


### Motifs

- Hard to find motif for *TAF1*. Is *TBP* a good motif to use?




## Cell types

```{r cellTypes}
'cell	lineage	tissue	karyotype	sex	description
GM12878	mesoderm 	blood 	normal 	F 	B-lymphocyte, lymphoblastoid, International HapMap Project - CEPH/Utah - European Caucasion, Epstein-Barr Virus
H1-hESC	inner cell mass 	embryonic stem cell 	normal 	M	embryonic stem cells
K562	mesoderm 	blood 	cancer 	F	leukemia, "The continuous cell line K-562 was established by Lozzio and Lozzio from the pleural effusion of a 53-year-old female with chronic myelogenous leukemia in terminal blast crises." - ATCC
Panc1		pancreas	cancer	M	pancreatic carcinoma, (PMID: 1140870) PANC-1 was established from a pancreatic carcinoma, which was extracted via pancreatico-duodenectomy specimen from a 56-year-old Caucasian individual. Malignancy of this cell line was verified via in vitro and in vivo assays.
IMR90	endoderm 	lung 	normal 	F	fetal lung fibroblasts, newly promoted to tier 2: not in 2011 analysis
PC-3		prostate	cancer	M	prostate cancer cell line
A549 	2 	epithelial cell line derived from a lung carcinoma tissue. (PMID: 175022), "This line was initiated in 1972 by D.J. Giard, et al. through explant culture of lung carcinomatous tissue from a 58-year-old caucasian male." - ATCC, newly promoted to tier 2: not in 2011 analysis 	endoderm 	epithelium 	cancer 	M 	Myers
Crawford
Stam   	ATCC CCL-185 	BTO:0000018 	A549
CD20+ 	2 	B cells from donors RO01778 and RO01794, newly promoted to tier 2: not in 2011 analysis 	mesoderm 	blood 	normal 	F 	Stam   	Dr. Heimfeld Lab at FHCRC CD20+ 	BTO:0000776 	B-cells CD20+
CD20+_RO01778 	2 	B cells, caucasian, draw number 1, newly promoted to tier 2: not in 2011 analysis 	mesoderm 	blood 	normal 	F 	Stam   	Rajinder Kaul, UW CD20+_RO01778 	BTO:0000776 	B-cells CD20+ (RO01778)
CD20+_RO01794 	2 	B cells, African American, draw number 1, newly promoted to tier 2: not in 2011 analysis 	mesoderm 	blood 	normal 	F 	Stam   	Rajinder Kaul, UW CD20+_RO01794 	BTO:0000776 	B cells CD20+ RO01794
H1-neurons 	2 	neurons derived from H1 embryonic stem cells, newly promoted to tier 2: not in 2011 analysis 	ectoderm 	neurons 	normal 	M 	Cellular Dynamics International Batch 1
Cellular Dynamics International Batch 2   	CDI TBD 	BTO:0000938 	H1-neurons
HeLa-S3 	2 	cervical carcinoma 	ectoderm 	cervix 	cancer 	F 	ENCODE
Crawford   	ATCC CCL-2.2 	BTO:0000568 	HeLa-S3
HepG2 	2 	hepatocellular carcinoma 	endoderm 	liver 	cancer 	M 	ENCODE   	ATCC HB-8065 	BTO:0000599 	HepG2
HUVEC 	2 	umbilical vein endothelial cells 	mesoderm 	blood vessel 	normal 	U 	ENCODE   	Lonza CC-2517 	BTO:0001949 	HUVEC
LHCN-M2 	2 	skeletal myoblasts derived from satellite cells from the pectoralis major muscle of a 41 year old caucasian heart transplant donor, immortalized with lox-hygro-hTERT ("LH"), and Cdk4-neo ("CN"), Zhu et al. (2007) in Aging Cell, vol. 6, pp 515-523, newly promoted to tier 2: not in 2011 analysis. 	mesoderm 	skeletal muscle myoblast 	  	M 	Wold   	Dr. Woodring Wright Missing 	BTO:0005046 	LHCN-M2
MCF-7 	2 	mammary gland, adenocarcinoma. (PMID: 4357757), newly promoted to tier 2: not in 2011 analysis 	ectoderm 	breast 	cancer 	F 	Crawford
Farnham
Stam   	ATCC HTB-22 	BTO:0000093 	MCF-7
Monocytes-CD14+ 	2 	Monocytes-CD14+ are CD14-positive cells from human leukapheresis production, from donor RO 01746 (draw 1 ID is RO 01746, draw 2 ID is RO 01826), newly promoted to tier 2: not in 2011 analysis 	mesoderm 	monocytes 	normal 	F 	Stam   	S. Heimfeld Laboratory, FHCRC Monocytes-CD14+ 	BTO:0000876 	Monocytes CD14+
Monocytes-CD14+_RO01746 	2 	Monocytes-CD14+ are CD14-positive cells from human leukapheresis production, from donor RO 01746 (draw 1 ID is RO 01746, draw 2 ID is RO 01826), Monocytes-CD14+_RO01746 and Monocytes-CD14+_RO01826 are being used as replicates, newly promoted to tier 2: not in 2011 analysis 	mesoderm 	monocytes 	normal 	F 	Stam   	S. Heimfeld Laboratory, FHCRC Monocytes-CD14+_RO01746 	BTO:0000876 	Monocytes CD14+ RO01746
Monocytes-CD14+_RO01826 	2 	Monocytes-CD14+ are CD14-positive cells from human leukapheresis production, Monocytes-CD14+_RO01826 is a second draw being used as a replicate for Monocytes-CD14+_RO01746, newly promoted to tier 2: not in 2011 analysis 	mesoderm 	monocytes 	normal 	F 	Stam   	S. Heimfeld Laboratory, FHCRC Monocytes-CD14+_RO01826 	BTO:0000876 	Monocytes CD14+ RO01826
SK-N-SH 	2 	neuroblastoma, newly promoted to tier 2: not in 2011 analysis, the SK-N-SH line was established from a neuroblastoma in a 4 year old individual. 	ectoderm 	brain 	cancer 	F 	Myers   	ATCC HTB-11 	BTO:0001620 	SK-N-SH

																														
																														
																														
																														
																														
																														
'
```
